src/hooks/useId.ts

```diff
@@ -1,9 +1,27 @@
 /* External dependencies */
 import { useMemo } from 'react'
-import { v4 as uuid } from 'uuid'
 
-function useId() {
-  return useMemo(() => uuid(), [])
+const id = Object.seal({ current: 0 })
+
+interface UseIdProps {
+  idProp?: string
+  prefix?: string
+}
+
+function useId({
+  idProp,
+  prefix,
+}: UseIdProps) {
+  return useMemo(() => {
+    if (idProp) { return idProp }
+    id.current += 1
+    return prefix
+      ? `${prefix}-${id.current}`
+      : `${id.current}`
+  }, [
+    idProp,
+    prefix,
+  ])
```

Comment:
note) useMemo를 semantic guarantee로서 사용하는 것은 지양해야 합니다. (이 동작은 useMemo에 넘기는 함수가 prefix, idProp이 바뀌지 않으면 단 한번 실행된다는 가정 하에 구현되어 있습니다.) memoization을 잊어버려 다시 실행하게 될 경우 id가 바뀌게 됩니다.

Reply:
좋은 구현은 아니라는 생각이 들었습니다. 나은 방식으로 변경해볼게요

Reply:
마땅한 방식이 떠오르진 않네요. https://github.com/chakra-ui/chakra-ui/blob/fa474bea3dcbdd4bbf2a26925f938d6e75a50c6d/packages/hooks/src/use-id.ts 참고해서 개선해봤습니다. 

useMemo를 사용한다는 점에서 기존과 크게 다르진 않은 거 같고, useEffect 내부에서 id를 증가시켜서 strict 모드일 때 id 증가가 두번 호출되는 걸 방지한다고 하네요

> We get the current id by context and generate a new id inside useEffect so that the side effects occur during the commit phase, Doing this prevents the side effects from being called twice when used with strict mode (render() in function component is the function body), which ends up making the server with the client not synchronized